pandas/tests/reshape/test_merge.py

```diff
@@ -1408,6 +1412,25 @@ def test_join_multi_dtypes(self, d1, d2):
         expected.sort_values(['k1', 'k2'], kind='mergesort', inplace=True)
         tm.assert_frame_equal(result, expected)
 
+    @pytest.mark.parametrize('int_vals, float_vals', [
+        ([1, 2, 3], [1.0, 2.0, 3.0]),
+        ([1, 2, 3], [1.0, 3.0]),
+        ([1, 2], [1.0, 2.0, 3.0]),
+        ([1, 2, 3], [1.1, 2.5, 3.0]),
+    ])
+    def test_merge_on_ints_floats(self, int_vals, float_vals):
+        # GH 16572
+        # Check that float column is not cast to object if
+        # merging on float and int
+        A = DataFrame({'X': int_vals})
+        B = DataFrame({'Y': float_vals})
+
+        res = A.merge(B, left_on='X', right_on='Y')
+        assert is_float_dtype(res['Y'].dtype)
```

Comment:
you need to check the results
the last merge is not what you would expect

Reply:
On my branch:
```
In [1]: A = pd.DataFrame({'X': [1, 2, 3]})
In [2]: B = pd.DataFrame({'Y': [1.1, 2.5, 3.0]})

In [3]: A.merge(B, left_on='X', right_on='Y'))
Out[3]: 
   X    Y
0  3  3.0

In [4]: A.merge(B, left_on='X', right_on='Y').dtypes
Out[4]:
X      int64
Y    float64
dtype: object
```
Is this not what we would expect? 

And could you give me an example of the problems caused when merging integer and float columns when the float values are not equal to the int representation of those same values? Thanks.

Reply:
hmm that looks ok, in any event compare the expected results (not the dtypes)

Reply:
the issue with comparing floats and integers is almost always wrong, IOW the user needs to be aware of this, as its almost always an error (so warning is ok here).